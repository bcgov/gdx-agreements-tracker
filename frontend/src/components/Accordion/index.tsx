import * as React from "react";
import { Accordion as MUIAccordion } from "@mui/material/";
import AccordionSummary from "@mui/material/AccordionSummary";
import AccordionDetails from "@mui/material/AccordionDetails";
import Typography from "@mui/material/Typography";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { styled } from "@mui/material";
import bcgovTheme from "../../bcgovTheme";

export const Accordion = ({
  children,
  sectionTitle,
}: {
  children: JSX.Element | JSX.Element[];
  sectionTitle: string;
}) => {
  const StyledAccordionSummary = styled(AccordionSummary)({
    backgroundColor: "#f3f3f3",
  });

  const StyledTypography = styled(Typography)({
    color: bcgovTheme.palette.primary.main,
    fontWeight: "bold",
  });

  return (
    <MUIAccordion defaultExpanded>
      <StyledAccordionSummary
        expandIcon={<ExpandMoreIcon />}
        aria-controls="panel-content"
        aria-label="button"
      >
        <StyledTypography>{sectionTitle}</StyledTypography>
      </StyledAccordionSummary>
      <AccordionDetails>{children}</AccordionDetails>
    </MUIAccordion>
  );
};
