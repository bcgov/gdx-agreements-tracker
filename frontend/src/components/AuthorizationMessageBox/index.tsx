import { FormHelperText } from "@mui/material";

export const AuthorizationMessageBox = ({ message }: { message: string }) => {
  return <FormHelperText>{message}</FormHelperText>;
};
