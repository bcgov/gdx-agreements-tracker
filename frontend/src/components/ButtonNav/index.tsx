import React from "react";
import { Link, useMatch, useResolvedPath } from "react-router-dom";
import type { LinkProps } from "react-router-dom";

/**
 * Returns react router link with active class if matches route.
 *
 * @param {LinkProps} root0 Link properties.
 * @param {ReactNode | undefined} root0.children Child nodes.
 * @param {To} root0.to Destination.
 * @returns {React.ReactNode}
 */
export const ButtonNav = ({ children, to, ...props }: LinkProps) => {
  const resolved = useResolvedPath(to);
  const match = useMatch({ path: resolved.pathname, end: true });
  if (match) {
    props.className = `${props.className ?? ""} active`;
  }
  return (
    <Link to={to} {...props}>
      {children}
    </Link>
  );
};

export default ButtonNav;
