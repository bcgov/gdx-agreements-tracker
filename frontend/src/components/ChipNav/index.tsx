import * as React from "react";
import { styled } from "@mui/material/styles";
import Chip from "@mui/material/Chip";
import { IChipNav } from "../../types";
import { Box, Paper } from "@mui/material";
import { NavLink } from "react-router-dom";
import bcgovTheme from "../../bcgovTheme";

export const ChipNav = ({
  navLinks,
  navLinksRight = [],
}: {
  navLinks: IChipNav[];
  navLinksRight?: IChipNav[];
}) => {
  const StyledListItem = styled("li")(({ theme }) => ({
    margin: theme.spacing(0.5),
  }));

  const StyledNavLink = styled(NavLink)(() => ({
    textDecoration: "none",
  }));

  /**
   * Renders a single ChipNav link.
   *
   * @param   {IChipNav}    link The ChipNav link to render.
   * @returns {JSX.Element}
   */
  const renderNavLink = (link: IChipNav): JSX.Element => {
    return (
      <StyledListItem key={link.key}>
        <StyledNavLink to={link.url} end>
          {({ isActive }) => (
            <Chip
              sx={{
                backgroundColor: isActive ? "#D3D3D3" : bcgovTheme.palette.primary.main,
                color: isActive ? "#000" : "#FFF",
              }}
              label={link.name}
            />
          )}
        </StyledNavLink>
      </StyledListItem>
    );
  };

  return (
    <Paper
      elevation={3}
      sx={{
        display: "flex",
        justifyContent: "space-between",
        flexWrap: "wrap",
        p: 0.5,
        mb: 2,
        mt: 0,
      }}
    >
      <Box sx={{ display: "flex", listStyle: "none" }}>{navLinks.map(renderNavLink)}</Box>
      {navLinksRight.length > 0 && (
        <Box sx={{ display: "flex", listStyle: "none" }}>{navLinksRight.map(renderNavLink)}</Box>
      )}
    </Paper>
  );
};
