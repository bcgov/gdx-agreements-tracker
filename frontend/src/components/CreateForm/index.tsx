import React from "react";
import { Box, Button } from "@mui/material";
import { Form, Formik, FormikHelpers, FormikValues } from "formik";
import { FormLayout } from "../GDXForm/FormLayout";
import { FormInput } from "../FormInput";
import { IEditFields } from "../../types";

/**
 * Destructured as:
 * @param    {​​​​​​Object[]}​​​​​​ initialValues - Contains all the initial values for Formik to use
 * @param    {​​​​​​​​​​​​​​​​​​​​string}​​​​​​​​​​​​​​​​​​​​ onSubmit - Handles the submit functionality for the form
 * @param    {Object[]}​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​ editFields - The field properties used to render the fields for an edit form
 * @returns  {React.JSX}
 */

export const CreateForm = ({
  initialValues,
  onSubmit,
  editFields,
}: {
  initialValues: FormikValues;
  // todo: Define a good type. "Any" type temporarily permitted.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onSubmit: ((values: unknown, formikHelpers: FormikHelpers<any>) => void | Promise<any>) &
    Function;
  editFields: IEditFields[];
}) => {
  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit}>
      {({ setFieldValue, values, handleChange, dirty }) => {
        return (
          <Form>
            <FormLayout>
              {editFields.map(({ fieldName, fieldType, fieldLabel, width, tableName }) => {
                return (
                  <FormInput
                    setFieldValue={setFieldValue}
                    fieldValue={values?.[fieldName]}
                    fieldName={fieldName}
                    fieldType={fieldType}
                    fieldLabel={fieldLabel}
                    handleChange={handleChange}
                    width={width}
                    key={fieldName}
                    tableName={tableName}
                  />
                );
              })}
            </FormLayout>
            <Box
              m={1}
              display="flex"
              justifyContent="flex-end"
              alignItems="flex-end"
              role={"submit_button"}
            >
              <Button
                type="submit"
                variant="contained"
                color="success"
                disabled={dirty ? false : true}
              >
                Submit
              </Button>
            </Box>
          </Form>
        );
      }}
    </Formik>
  );
};
