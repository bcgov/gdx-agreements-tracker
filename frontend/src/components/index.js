export { ButtonNav } from "./ButtonNav";
export { Debug } from "./Debug";
export { Loader } from "./Loader";
export { Main } from "./Layout/Main";
export { PageLink } from "./PageLink";
export { RowEditDeleteChip } from "./RowEditDeleteChip";
export { Sidebar } from "./Sidebar";
export { DetailsList } from "./DetailsList";
export { Table } from "./Table";
