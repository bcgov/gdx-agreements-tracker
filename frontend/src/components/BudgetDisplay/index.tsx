import { Box, Grid } from "@mui/material";
import { GDXList } from "components/GDXList";
import { Loader } from "components/Loader";
import { useAxios } from "hooks/useAxios";
import React from "react";
import { useQuery } from "react-query";

export const BudgetDisplay = ({ apiUrl }: { apiUrl: string }) => {
  const { axiosAll } = useAxios();

  const getBudgetData = async () => {
    const budget = await axiosAll()
      .get(apiUrl)
      .then((budgetData: any) => {
        return budgetData;
      });
    return budget.data.data;
  };

  // Queries
  const { data, isLoading } = useQuery("contract budget", getBudgetData, {
    // todo: When there is an edit and view form built, reassess these options.
    refetchOnWindowFocus: false,
    retryOnMount: false,
    refetchOnReconnect: false,
    retry: false,
    staleTime: Infinity,
  });

  return (
    <Grid container spacing={2}>
      {!data ? (
        <Loader />
      ) : (
        data.map((budget: any) => {
          return <GDXList data={budget} title={`Fiscal ${budget.fiscal_year}`} />;
        })
      )}
    </Grid>
  );
};