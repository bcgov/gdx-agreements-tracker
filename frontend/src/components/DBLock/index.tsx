import { Grid, Skeleton } from "@mui/material";
import LockPersonIcon from "@mui/icons-material/LockPerson";
import { FormEditButton } from "components/FormEditButton";
import { useParams } from "react-router";

interface IDBLockProps {
  handleDbLock: Function;
  removeLock: Function;
  // todo: Define a good type. "Any" type temporarily permitted.  The query:any type will be fixed when we upgrade to ReactQuery V4.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  query: any;
  handleFormType: (formType: "read" | "edit" | "new") => void;
}

/* This is a functional component called `DBLock` that takes in four props: `handleDbLock`,
`removeLock`, `query`, and `handleEditMode`. It returns a JSX element that renders a grid with
various components, including `Skeleton`, `LockPersonIcon`, and `FormEditButton`. The `query` prop
is of type `any` for now, but will be fixed when the code is upgraded to ReactQuery V4. The
component also uses the `useParams` hook from React Router to get the `projectId` parameter from the
URL. When the `FormEditButton` is clicked, it calls various functions passed in as props to remove a
lock on a database row, handle a new lock, and set the edit mode to true. */

export const DBLock = ({ handleDbLock, removeLock, query, handleFormType }: IDBLockProps) => {
  const { projectId } = useParams();

  return (
    <Grid container spacing={2}>
      <Grid item xs={6} md={6}>
        <Skeleton variant="rectangular" />
      </Grid>
      <Grid item xs={6} md={6}>
        <Skeleton variant="rectangular" />
      </Grid>
      <Grid item xs={6} md={6}>
        <LockPersonIcon />
        {/* <h1>section locked for editing by: {query?.data?.dbRowLock.locked_by}</h1> */}
      </Grid>
      <Grid item xs={6} md={6}></Grid>
      <Grid item xs={4} md={4}></Grid>
      <Grid item xs={4} md={4}></Grid>
      <Grid item xs={4} md={4}>
        <FormEditButton
          buttonText="Take Over Editing"
          onClick={async () => {
            await removeLock(query?.data?.dbRowLock).then(async () => {
              await handleDbLock(query, projectId).then(async () => {
                await query.refetch().then(() => {
                  handleFormType("edit");
                });
              });
            });
          }}
        />
      </Grid>
    </Grid>
  );
};
