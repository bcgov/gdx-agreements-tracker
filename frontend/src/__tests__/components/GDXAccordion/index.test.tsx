import React from "react";
import { GDXAccordion } from "../../../components";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

describe("Layout GDXAccordion", () => {
  it("Renders at all.", () => {
    const { container } = render(
      <GDXAccordion sectionTitle={"testSection"}>
        <div>testChild component</div>
      </GDXAccordion>,
      {
        wrapper: MemoryRouter,
      }
    );
    expect(container).not.toBeEmptyDOMElement();
  });
});
