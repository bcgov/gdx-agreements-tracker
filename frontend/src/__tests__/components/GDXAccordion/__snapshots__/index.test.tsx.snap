// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests different renders of the <Renderer /> component test 1 1`] = `"<div class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root\\"><div class=\\"MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-14pd8st-MuiButtonBase-root-MuiAccordionSummary-root\\" tabindex=\\"0\\" role=\\"button\\" aria-expanded=\\"true\\" aria-controls=\\"panel-content\\" aria-label=\\"button\\"><div class=\\"MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content\\"><p class=\\"MuiTypography-root MuiTypography-body1 css-1ug9r71-MuiTypography-root\\">mock title</p></div><div class=\\"MuiAccordionSummary-expandIconWrapper Mui-expanded css-yw020d-MuiAccordionSummary-expandIconWrapper\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root\\" focusable=\\"false\\" aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" data-testid=\\"ExpandMoreIcon\\"><path d=\\"M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z\\"></path></svg></div></div><div class=\\"MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root\\" style=\\"min-height: 0px;\\"><div class=\\"MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper\\"><div class=\\"MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner\\"><div id=\\"panel-content\\" role=\\"region\\" class=\\"MuiAccordion-region\\"><div class=\\"MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root\\"><div>mock component</div></div></div></div></div></div></div>"`;
